File: app/src/main/groovy/us/kesslern/ascient/receiver/SMSBroadcastReceiver.java
package us.kesslern.ascient.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.SmsMessage;
import android.util.Log;

import org.codehaus.groovy.runtime.DefaultGroovyMethods;
import org.springframework.web.client.RestTemplate;

import us.kesslern.ascient.activity.MainActivity;

public class SMSBroadcastReceiver extends BroadcastReceiver {
    private MainActivity activity;
    private RestTemplate restTemplate;

    @Override
    public void onReceive(Context context, Intent intent) {
        Log.v("test", "received");

        Bundle bundle = intent.getExtras();

        if (DefaultGroovyMethods.asBoolean(bundle)) {
            Object[] pdus = DefaultGroovyMethods.asType(DefaultGroovyMethods.getAt(bundle, "pdus"), Object[].class);

            for (Object pdu : pdus) {
                SmsMessage smsMessage = SmsMessage.createFromPdu((byte[]) pdu);
                Log.v("test", smsMessage.getDisplayMessageBody());
                activity.update();
            }
        }
    }

    public MainActivity getActivity() {
        return activity;
    }

    public void setActivity(MainActivity activity) {
        this.activity = activity;
    }

    public RestTemplate getRestTemplate() {
        return restTemplate;
    }

    public void setRestTemplate(RestTemplate restTemplate) {
        this.restTemplate = restTemplate;
    }
}

File: app/src/main/java/us/kesslern/ascient/SMSBroadcastReceiver.kt
package us.kesslern.ascient

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.telephony.SmsMessage
import android.util.Log

import android.provider.Telephony
import android.os.Build


class SMSBroadcastReceiver(private val activity: MainActivity) : BroadcastReceiver() {

    private val TAG = tag(SMSBroadcastReceiver::class)

    override fun onReceive(context: Context, intent: Intent) {
        Log.v("test", "received")

        val bundle = intent.extras
        val smsMessage: SmsMessage

        smsMessage =
                if (Build.VERSION.SDK_INT >= 19) {
                    val msgs = Telephony.Sms.Intents.getMessagesFromIntent(intent)
                    msgs[0]
                } else {
                    val pdus = bundle.get("pdus") as Array<*>
                    SmsMessage.createFromPdu(pdus[0] as ByteArray)
                }

        Log.v(TAG, smsMessage.displayMessageBody)
        activity.update()
    }
}
